Step1 pr 보냅니다.

jar100 opened the pull request:
# 블랙잭 게임

## 구현 완료
1. 기본 룰 구현

2. 버스트 구현

3. user hti/ stand 선택 구현

4. 승패확인을 map으로 구현

5. a의 값을 1/11로 상황에 맞게 구현


## 추후 구현예정
1. 리팩토링

2. 승패 결과 다시 리팩토링 중복이 너무 많아보임

3. 다 대 일로 구현

4. 웹 구현

5 . 칩 구현


## 포피에게 하고싶으말 
많은 피드백 부탁드려용

jar100 committed:
player, card, cardDeck 객체 생성

#todo
- 카드 객체안에 있는 숫자를 객체화
- 플레이어가 카드를 가질 수 있도록
- 딜러 객체 생성

jar100 committed:
step1 card enum 구현

#todo
- 카드댁 픽업
- 카드댁 셔플
- 플래이어와 카드 연관관계
- 딜러 객체 생성

jar100 committed:
step1 카드덱과 플레이어 연관관계 구현

#todo
- 게임 생성
- 인풋 아웃풋 구현

jar100 committed:
step1 중간 완성

# 구현
- 유저가 히트와 스탠드를 선택해 카드를 받을 수 있게 구현

# todo
- 유저가 스텐드 하면 딜러의 카드를 16 초과가 될 때 까지 카드를 받게 구현
- 버스트 구현 해야한다.
- 총합 출력

kkssry committed:
step1 Player 버스트 유무 구현

#todo
A의 default값은 11인데  Total score가 21에 근접일 경우에 따라 1일지 11일지 판단
User 칩 관리(승패 여부에 따른 칩의 변화)
출력부를 View layer로 이동하여 기능 세분화

kkssry committed:
Merge pull request #1 from kkssry/step1

step1 Player 버스트 유무 구현

jar100 committed:
step1 2/12 중간커밋

a 가 1혹은  11 로 전환구현
기타 잔기능과 리팩토링, 티디디 구현

jar100 committed:
step1 2/12 마지막 커밋

# 구현완료
- a 를 1과 11 로 둘다 출력 가능하게 구현 (리펙토링 필요)
- 버스트 일부 구현

#todo
- 유저가 버스트 나올시 딜러가 승리하는 조건 구현
- 유저 블랙잭 나올시 1.5배 받고 게임종료구현
- 침 관리 구현

kkssry committed:
step1 2/14 User가 burst나올시 dealer가 승리하는 조건 구현

#구현완료
- 게임의 승리 조건 구현(User가이길경우, dealer가 이길경우, 비기는 경우, burst(21초과) 상황 구현
- 카드 문양에 그림을 삽입

#todo
- refactoring : BlackjackGame클래스에서 Result인스턴스 변수를  enum으로 구현해 조건문(if) 중복 제거
- 칩 관리 구현

kkssry committed:
step1 refactoring 진행중

kkssry committed:
step1 2/14 구현 완료 후 리팩토링중

#todo
- 유저턴을 유저가 가지게 리팩토링
- checkBlackjack 과 winner 메소드의 중복제거
- 칩 구현

kkssry committed:
Merge pull request #2 from kkssry/step1

Step1

jar100 committed:
step1 2/15 블랙잭 리팩토링

checkblackjack의 if 를 재거했다.

jar100 committed:
승패여부 메서드에서 map을 통해 if문 제거

jar100 committed:
step1 refactoring

javajigi commented on src/main/java/javablackjack/blackjack/domain/BlackjackGame.java:
```diff
@@ -0,0 +1,105 @@
+        checkResultCase(resultCases.blackjackCase());
+
+        System.out.println("딜러의 총합 : " + dealer.score());
+        System.out.println("유저의 총합 : " + user.score());
```
logging 라이브러리로 구현할 것을 추천

javajigi commented on src/main/java/javablackjack/blackjack/domain/BlackjackGame.java:
```diff
@@ -0,0 +1,105 @@
+        return result.getGameResult();
+    }
+
+    private void checkResultCase(Map<Integer, BiFunction<Player, Player, GameResult>> cases) {
```
user, dealer를 활용해 GameResult를 구할 수는 없나?
`GameResult result = dealer.result(user);`와 같은 구현은 힘을까?

javajigi commented on src/main/java/javablackjack/blackjack/domain/BlackjackGame.java:
```diff
@@ -0,0 +1,105 @@
+    //todo : 리팩토링 해야함
+    public void dealerTurn() {
+        if (result.equals(GameResult.DEAULF)) {
+            while (dealer.score() < 17) {
```
하드코딩하지 마라.

javajigi commented on src/main/java/javablackjack/blackjack/domain/BlackjackGame.java:
```diff
@@ -0,0 +1,105 @@
+    // todo 이걸 밖으로 빼면 렌덤 태스트가 가능하다. 의존관계 분리
+    private CardDeck cardDeck = new CardDeck();
+
+    private GameResult result = GameResult.DEAULF;
```
굳이 인스턴스 변수로 구현해야 하나?
로컬 변수로 구현할 수는 없을까?

javajigi commented on src/main/java/javablackjack/blackjack/domain/Player.java:
```diff
@@ -0,0 +1,79 @@
+    //todo : 1급 객체로 리팩토링 해야함.
+    private List<Card> cards = new ArrayList<>();
+    private String name;
+    private boolean playerTurn = true;
```
이 인스턴스 변수가 굳이 필요한가?

javajigi commented on src/main/java/javablackjack/blackjack/domain/ResultCases.java:
```diff
@@ -0,0 +1,40 @@
+
+public class ResultCases {
+
+    public Map<Integer, BiFunction<Player, Player, GameResult>> burstCase() {
```
`Map<Integer, BiFunction<Player, Player, GameResult>>` 구조가 너무 복잡한데 좀 단순하게 구현할 수는 없을까?
Map이 아닌 List를 사용할 수는 없나? 반드시 BiFunction을 써야 하나?
`Player(user), Player(dealer)` 두 개를 인스턴스 변수로 가지는 Pair와 같은 객체를 추가해 해결할 수는 없나?

javajigi reviewed (changes requested):
객체 분리나 전체적인 구현, 테스트 구현 등 깔끔하게 잘 구현하고 있네요. 👍

kkssry committed:
step1 2/18 feedback반영

kkssry committed:
Merge pull request #4 from kkssry/step1

step1 2/18 feedback반영

jar100 committed:
step1 02/19

# 구현
resultCases 클래스 인터페이스로 추출
cardDeckFactory 클래스 구현
bifunction 을 function 으로 리팩토링

# todo
칩 구현
펑션 메소드 리팩토링

kkssry committed:
step1 2019/2/20 칩 관리 구현

#todo
칩 관리 리팩토링
resultCase 리팩토링

kkssry committed:
step1 2019/2/21 chip re-push

kkssry committed:
Merge pull request #6 from kkssry/step1

Step1

jar100 committed:
step1 리팩토링

jar100 committed:
step1 리팩토링2

jar100 committed:
step1 2019/2/21 완료!

javajigi reviewed (approved):
테스트까지 고려하면서 구현 잘 했네. 💯 
다음 단계는 추가 기능 구현보다는 이 기능을 웹 버전으로 구현할 것을 추천한다.
